src/lib.rs

```diff
@@ -287,69 +327,518 @@ impl<D: AsMut<[Digit]>> BigInt<D> {
+    ///
+    /// ## Safety
+    /// Assuming that `a` and `b` have [sizes][size] `n` and `m` respectively:
+    /// * `n >= m > 0`
+    /// * [Size][size] of `self` must be exactly `n-m` limbs
+    /// * Size of `scratch_space` must be exactly
+    /// [`ScratchSpace::division(n, m)`](ScratchSpace::division) limbs
+    ///
+    /// Note: we check that these requirements are met using debug assertions.
+    ///
+    /// [size]: BigInt::size
+    ///
+    /// ## Constant time
+    /// This function is constant time in size of its arguments, ie. number of instructions and
+    /// memory access patterns remain the same for the same `n` and `m`.
+    pub unsafe fn div_unchecked(
+        &mut self,
+        a: &mut BigInt<impl AsMut<[Digit]>>,
+        b: &BigInt<impl AsRef<[Digit]>>,
+        mut scratch_space: ScratchSpace,
+    ) {
+        let n = a.as_mut().len() - b.size();
+        let carry = ops::div_n_m(
+            a.as_mut(),
+            b.as_ref(),
+            scratch_space.as_mut(),
+            &mut self.as_mut()[..n],
+        );
+        self.as_mut()[n] = carry;
     }
 }
 
-impl<'n, D: AsRef<[Digit]>> Add<&'n BigInt<D>> for u8 {
+impl<D: AsRef<[Digit]>> Div<Digit> for &BigInt<D> {
     type Output = BigInt;
 
-    #[inline(always)]
-    fn add(self, rhs: &BigInt<D>) -> Self::Output {
-        rhs + self
+    /// Computes `self / rhs` where `rhs` is a `Digit`.
+    ///
+    /// ## Constant time
+    /// Irrespective of it's inputs, this method takes `O(self.size())` steps.
+    ///
+    /// ## Zeroizing
+    /// This function might allocate some auxiliary buffer on the heap.
+    /// Before deallocating, it gets zeroed out.
+    ///
+    /// ## Complexity
+    /// The complexity is `O(self.size())`, which is asymptotically optimal.
+    ///
+    /// ## Panics
+    /// The method panics if `rhs` equals zero.
+    fn div(self, rhs: Digit) -> Self::Output {
+        self / &BigInt::from_digits(&[rhs])
+    }
+}
+
+impl<D: AsMut<[Digit]>> BigInt<D> {
+    /// Computes `floor(a / b)` where `b` is a digit, writes the result to `self`.
+    ///
+    /// It is a low-level function allowing you to manage allocations. Its safe analogue is
+    /// [`&a / b`](#impl-Div%3CDigit%3E).
+    ///
+    /// ## Safety
+    /// Assuming that `a` has [size] `n`:
+    /// * `n > 0`
+    /// * [Size][size] of `self` must be exactly `n` limbs
+    /// * Size of `scratch_space` must be exactly
+    /// [`ScratchSpace::division(n, 1)`](ScratchSpace::division) limbs
+    ///
+    /// [size]: BigInt::size
+    ///
+    /// ## Constant time
+    /// This function is constant time in size of its arguments, ie. number of instructions and
+    /// memory access patterns remain the same for the same `n`.
+    pub unsafe fn div_digit_unchecked(
+        &mut self,
+        a: &mut BigInt<impl AsMut<[Digit]>>,
+        b: Digit,
+        mut scratch_space: ScratchSpace,
+    ) {
+        self.div_unchecked(a, &BigInt::from_digits(&[b]), scratch_space);
```

Comment:
The same as for `mul_digit_unchecked` — having this function may be confusing. We need either to remove it or to document that there's no performance difference.

Reply:
Also, I'm not sure how different this is from addition